commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Listener definitions.
+ *
+ * DESIGN RATIONALE:
+ * <ul>
+ *    <li>The current set of listener events that can be fired are related
+ *    to modifications, and deciding between create or modify, or providing
+ *    removed entry information, require reading the previous value. Hence,
+ *    the current set of listeners can only be provided for read-write maps.
+ *    In the future, if for example cache entry visited events are to be
+ *    supported, those would need to be associated with either a read or
+ *    read-write map, and hence the listeners interface would most likely
+ *    be split up.
```

Comment:
Listeners mean interests. You have interests for operations that you've initiated and that others initiate. Say you initially only have interests for write-only events, but someone else executes a read-write operation, then you won't receive the event because it's not a write-only event. If you want to listen for read-write operations, you need a listener there too.

One could argue: A read-write operation does a write, why not receive a write-only event with the written data? Because it's not the same thing. If you want to listen for read-write events, implement such a listener.

Another person might argue: A write-only operation does a write, why not receive a read-write event with before being null or Optional.empty (if using optionals)? Because a write-only operation is not the same as a read-write one and adding optionals/null here would be a mess. A read-write operation listener guarantees that it'll give you the previous value and the updated value.